### PR TITLE
refactor workspace controller

### PR DESCRIPTION
**Pull request by wansir**
Signed-off-by: hongming <talonwan@yunify.com>

**What type of PR is this?**

 /kind cleanup

**What this PR does / why we need it**:

1. Refactor workspace controller
2. Refactor workspace role controller
3. Refactor workspace template controller
4. Refactor namespace controller
5. Add some test code.

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: pkg/controller/workspacetemplate/workspacetemplate_controller.go [hunk 1]
```diff
@@ -91,128 +70,52 @@ func NewController(k8sClient kubernetes.Interface, ksClient kubesphere.Interface
...
-	}(obj)
-
-	if err != nil {
-		utilruntime.HandleError(err)
-		return true
-	}
-
-	return true
+func (c *workspaceTemplateController) Run(threadiness int, stopCh <-chan struct{}) error {
+	return c.Run(5, stopCh)
```
Comment (RolandMa1986) [inline pkg/controller/workspacetemplate/workspacetemplate_controller.go:112]: Run function should be removed. This will cause a cascade call.

**Comment by wansir** [comment]
I'll  add more test code.

**Comment by wansir** [comment]
resource-hierarchy

File: pkg/apis/types/v1beta1/federatedworkspacerole_types.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ResourcePluralFederatedWorkspaceRole   = "federatedworkspaceroles"
+	ResourceSingularFederatedWorkspaceRole = "federatedworkspacerole"
+	FederatedWorkspaceRoleKind             = "FederatedWorkspaceRole"
+)
+
+// +genclient
```
Comment (RolandMa1986) [inline pkg/apis/types/v1beta1/federatedworkspacerole_types.go:30]: `+genclient` could be deleted when using Contoller-runtime. I don't think the auto-genated clients were used anywhere.

File: cmd/controller-manager/app/controllers.go [hunk 3]
```diff
@@ -185,27 +180,14 @@ func addControllers(
...
 		fedGlobalRoleCache, fedGlobalRoleCacheController = util.NewResourceInformer(fedGlobalRoleClient, "", &iamv1alpha2.FedGlobalRoleResource, func(object runtime.Object) {})
 		fedGlobalRoleBindingCache, fedGlobalRoleBindingCacheController = util.NewResourceInformer(fedGlobalRoleBindingClient, "", &iamv1alpha2.FedGlobalRoleBindingResource, func(object runtime.Object) {})
-		fedWorkspaceRoleCache, fedWorkspaceRoleCacheController = util.NewResourceInformer(fedWorkspaceRoleClient, "", &iamv1alpha2.FedWorkspaceRoleResource, func(object runtime.Object) {})
-		fedWorkspaceRoleBindingCache, fedWorkspaceRoleBindingCacheController = util.NewResourceInformer(fedWorkspaceRoleBindingClient, "", &iamv1alpha2.FedWorkspaceRoleBindingResource, func(object runtime.Object) {})
 
 		go fedUserCacheController.Run(stopCh)
 		go fedGlobalRoleCacheController.Run(stopCh)
 		go fedGlobalRoleBindingCacheController.Run(stopCh)
-		go fedWorkspaceRoleCacheController.Run(stopCh)
 		go fedWorkspaceRoleBindingCacheController.Run(stopCh)
```
[the remark this answers is not in the document]
Comment (wansir) [reply cmd/controller-manager/app/controllers.go]: done

File: pkg/apis/types/v1beta1/federatedworkspacerole_types.go (hunk 2 above)
Comment (wansir) [reply pkg/apis/types/v1beta1/federatedworkspacerole_types.go]: done

File: pkg/controller/workspacetemplate/workspacetemplate_controller.go [hunk 4]
```diff
@@ -18,387 +18,298 @@ package workspacetemplate
...
 	}
-	for _, roleBase := range roleBases {
-		var role iamv1alpha2.WorkspaceRole
-		if err = yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer(roleBase.Role.Raw), 1024).Decode(&role); err == nil && role.Kind == iamv1alpha2.ResourceKindWorkspaceRole {
-			roleName := fmt.Sprintf("%s-%s", workspace.Name, role.Name)
-			if role.Labels == nil {
-				role.Labels = make(map[string]string, 0)
+	for _, template := range templates.Items {
+		var expected iamv1alpha2.WorkspaceRole
+		if err := yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer(template.Role.Raw), 1024).Decode(&expected); err == nil && expected.Kind == iamv1alpha2.ResourceKindWorkspaceRole {
```
Comment (RolandMa1986) [inline pkg/controller/workspacetemplate/workspacetemplate_controller.go:283]: The `err` may need to be logged properly

File: pkg/controller/namespace/namespace_controller.go [hunk 5]
```diff
@@ -263,67 +234,73 @@ func (r *ReconcileNamespace) initRoles(namespace *corev1.Namespace) error {
...
-				if err := r.Update(context.Background(), &old); err != nil {
-					klog.Error(err)
+				logger.V(4).Info("update builtin role", "role", template.Name)
+				if err := r.Update(ctx, &old); err != nil {
+					logger.Error(err, "update role failed")
 					return err
 				}
 			}
+		} else if err != nil {
+			logger.V(4).Info("invalid role base found", "name", roleBase.Name, "err", err.Error())
```
Comment (RolandMa1986) [inline pkg/controller/namespace/namespace_controller.go:271]: The Log level should be increased. Please ref to https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md

**Review by RolandMa1986 — COMMENTED** [review]
Other changes LGTM.

File: pkg/controller/workspacetemplate/workspacetemplate_controller.go (hunk 4 above)
Comment (wansir) [reply pkg/controller/workspacetemplate/workspacetemplate_controller.go:239]: done

File: pkg/controller/namespace/namespace_controller.go (hunk 5 above)
Comment (wansir) [reply pkg/controller/namespace/namespace_controller.go]: done

**Comment by wansir** [comment]
Add unbind workspace error check in namespace controller.

File: cmd/controller-manager/app/server.go [hunk 6]
```diff
@@ -186,13 +189,29 @@ func run(s *options.KubeSphereControllerManagerOptions, stopCh <-chan struct{})
 		klog.Fatalf("unable add APIs to scheme: %v", err)
 	}
 
-	err = workspace.Add(mgr)
-	if err != nil {
+	if err = (&workspacetemplate.Reconciler{
+		MultiClusterEnabled: s.MultiClusterOptions.Enable,
+	}).SetupWithManager(mgr); err != nil {
+		klog.Fatal("Unable to create workspace template controller")
```
[the remark this answers is not in the document]
Comment (wansir) [reply cmd/controller-manager/app/server.go:196]: fixed

File: cmd/controller-manager/app/server.go [hunk 7]
```diff
@@ -186,13 +189,29 @@ func run(s *options.KubeSphereControllerManagerOptions, stopCh <-chan struct{})
...
 
-	err = namespace.Add(mgr)
-	if err != nil {
+	if err = (&workspacerole.Reconciler{
+		MultiClusterEnabled: s.MultiClusterOptions.Enable,
+	}).SetupWithManager(mgr); err != nil {
+		klog.Fatal("Unable to create workspace role controller")
+	}
+
+	if err = (&workspacerolebinding.Reconciler{
```
[the remark this answers is not in the document]
Comment (wansir) [reply cmd/controller-manager/app/server.go]: fixed

**Pull request merged**